src/exporters/bristol.rs

```diff
@@ -1,100 +1,117 @@
+                } else {
+                    writeln!(sink, "1 1 {} {} EQW", i, o) // identity gate
+                }
+            }
+            Operation::Sub(o, l, r) => {
+                writeln!(sink, "2 1 {} {} {} XOR", l, r, o) // ADD and SUB are equivalent on GF2
+            }
+            Operation::SubConst(o, i, c) => {
+                if *c {
+                    writeln!(sink, "1 1 {} {} INV", i, o)
+                } else {
+                    writeln!(sink, "1 1 {} {} EQW", i, o) // identity gate
+                }
+            }
+            Operation::Mul(o, l, r) => {
+                writeln!(sink, "2 1 {} {} {} AND", l, r, o)
+            }
+            Operation::MulConst(o, i, c) => {
+                if *c {
+                    writeln!(sink, "1 1 {} {} EQW", i, o) // identity gate
+                } else {
+                    writeln!(sink, "1 1 0 {} EQ", o)
+                }
+            }
+            Operation::AssertZero(w) => {
+                // Bristol doesn't really have a concept of output wires _or_ assertions, so this
+                // non-spec representation is the best we can do.
+                writeln!(sink, "0 1 {} OUTPUT", w)
+            }
+            Operation::Const(w, c) => {
+                writeln!(sink, "1 1 {} {} EQ", if *c { 1 } else { 0 }, w)
             }
-        }
-        Operation::AssertZero(w) => {
-            // Bristol doesn't really have a concept of output wires _or_ assertions, so this
-            // non-spec representation is the best we can do.
-            format!("0 1 {} OUTPUT", w)
-        }
-        Operation::Const(w, c) => {
-            format!("1 1 {} {} EQ", if *c { 1 } else { 0 }, w)
         }
     }
-}
 
-pub fn bool_circuit_to_bristol(gates: &[Operation<bool>], bool_witness: &[bool]) -> String {
-    let mut circuit: String = String::new();
-    let mut bool_iter = bool_witness.iter();
+    fn export_circuit(
+        gates: &[Operation<bool>],
+        witness: &[bool],
+        sink: &mut impl Write,
+    ) -> Result<()> {
+        let mut wit_iter = witness.iter();
 
-    for gate in gates {
-        circuit.push_str(
-            format!(
-                "{}\n",
-                match gate {
-                    Operation::Input(o) => {
-                        bool_gate_to_bristol(&Operation::Const(*o, *bool_iter.next().unwrap()))
-                    }
-                    _ => {
-                        bool_gate_to_bristol(gate)
-                    }
-                }
-            )
-            .as_str(),
-        )
-    }
+        for gate in gates {
+            match gate {
+                Operation::Input(o) => Self::export_gate(
+                    &Operation::Const(
+                        *o,
+                        *wit_iter
+                            .next()
+                            .ok_or_else(|| Error::new(ErrorKind::Other, "witness too short"))?,
+                    ),
```

Comment:
Yeah, I think that would be ideal. I want to save that for a later refactor, though, since I'm still familiarizing myself with this codebase.

(If I understand correctly, we should always have exactly as many witness inputs as there are `Input` gates, so making these covariant is always correct for our usecase.)